src/node_contextify.cc

```diff
@@ -1470,19 +1489,61 @@ void ContextifyContext::ContainsModuleSyntax(
                                                    id_symbol,
                                                    try_catch);
 
-  bool found_error_message_caused_by_module_syntax = false;
+  bool should_retry_as_esm = false;
   if (try_catch.HasCaught() && !try_catch.HasTerminated()) {
     Utf8Value message_value(env->isolate(), try_catch.Message()->Get());
     auto message = message_value.ToStringView();
 
     for (const auto& error_message : esm_syntax_error_messages) {
       if (message.find(error_message) != std::string_view::npos) {
-        found_error_message_caused_by_module_syntax = true;
+        should_retry_as_esm = true;
+        break;
+      }
+    }
+
+    for (const auto& error_message : throws_only_in_cjs_error_messages) {
+      if (message.find(error_message) != std::string_view::npos) {
+        // Try parsing again where the user's code is wrapped within an async
+        // function. If the new parse succeeds, then the error was caused by
+        // either a top-level declaration of one of the CommonJS module
+        // variables, or a top-level `await`.
+        TryCatchScope second_parse_try_catch(env);
+        Local<String> wrapped_code =
+            String::Concat(isolate,
+                           String::NewFromUtf8(isolate, "(async function() {")
+                               .ToLocalChecked(),
+                           code);
+        wrapped_code = String::Concat(
+            isolate,
+            wrapped_code,
+            String::NewFromUtf8(isolate, "})();").ToLocalChecked());
+        ScriptCompiler::Source wrapped_source =
+            GetCommonJSSourceInstance(isolate,
```

Comment:
Instead of wrapping with both a string and the CJS wrapper, could we instead pass the `code` to the `AsyncFunction` constructor?

Reply:
Wouldn't doing that fix the TLA case but not the redeclared variables case?

Reply:
Do you have an example of how to use the `AsyncFunction` constructor? I’m not finding any usages in the codebase where it seems obvious how I’d use it here.

Re “wrapping” the string, all I’m doing is taking the original source code string and creating a new one via concatenation. I can’t imagine that string concatenation causes a noticeable performance impact? Likewise is `AsyncFunction` noticeably faster than the CommonJS wrapper?

I’d rather not hold up a bugfix on a performance optimization, especially since all of this code is for an infrequent code path; it’s not for all usages of top-level `await` or `const require`, but instances of those _above the first `import` or `export`._ I’m not opposed to the optimization, of course, but perhaps it can come as a follow-up?

Reply:
Let’s say the user’s source is:

```js
const fs = require('fs');
await fs.readFile('./package.json');
```

The first parse (as CommonJS) would throw on the top-level `await`. Per this PR, then we would do a second parse as CommonJS where the original source is wrapped by `(async function() {` and `})();`, which would parse without error, and therefore we retry as ESM which fails on undefined `require`, and this is the error reported to the user. This is all correct per the docs, since top-level `await` is one of the syntaxes that should trigger running as ESM, and of course `require` is undefined in ESM.

Let’s say I changed the logic for the second parse to replace the CommonJS wrapper with just an async function. Now the second parse throws on undefined `require`, and we stop. In this PR whenever the two CommonJS parses both throw errors, the error that we return is the error from the first parse, since the second one might be nonsense since the second parse wasn’t of the user’s code. So in this case the user gets the first parse error, about unsupported top-level `await`, and that’s the wrong error per the docs.

If we replace the CommonJS wrapper with an _async_ CommonJS wrapper, i.e. instead of `(function(module, exports, require, __filename, __dirname) {` we parse `(async function(module, exports, require, __filename, __dirname) {`, then we avoid the “undefined `require`“ problem in the second parse but we still have the other case that this PR is meant to fix, where the user should be able to write `const require` in ESM code. By providing the CommonJS variables we prevent the undefined `require` but _cause_ the variable redeclaration problem.

So TLDR is I think using the `AsyncFunction` constructor won’t work, but I don’t feel all that confident about it if you think you can show that I’m wrong 😄

Reply:
The `AsyncFunction` idea won’t work for many reasons, the main one is that V8 doesn’t expose any API I could find to do it.